connman/vpn/plugins/wireguard.c

```diff
@@ -293,42 +298,89 @@ static bool sockaddr_cmp_addr(struct sockaddr_u *a, struct sockaddr_u *b)
 
 static void run_dns_reresolve(struct wireguard_info *info);
 
-static gboolean wg_dns_reresolve_cb(gpointer user_data)
+struct resolve_data {
+	struct wireguard_info *info;
+	struct sockaddr_u *addr;
+};
+
+static gboolean remove_resolv(gpointer user_data)
 {
 	struct wireguard_info *info = user_data;
-	struct sockaddr_u addr;
+
+	DBG("");
+
+	if (info->remove_resolv_id)
+		g_source_remove(info->remove_resolv_id);
+
+	if (info->resolv && info->resolv_id) {
+		DBG("cancel resolv lookup");
+		vpn_util_cancel_resolve(info->resolv, info->resolv_id);
+	}
+
+	info->resolv_id = 0;
+	info->remove_resolv_id = 0;
+
+	vpn_util_resolve_unref(info->resolv);
+	info->resolv = NULL;
+
+	return G_SOURCE_REMOVE;
+}
+
+static void resolve_endpoint_cb(GResolvResultStatus status,
+					char **results, gpointer user_data)
+{
+	struct resolve_data *data = user_data;
+	struct wireguard_info *info = data->info;
 	int err;
 
-	if (vpn_provider_get_connection_errors(info->provider) >=
-						DNS_RERESOLVE_ERROR_LIMIT) {
-		connman_warn("reresolve error limit reached");
-		vpn_died(NULL, -ENONET, info->provider);
-		return G_SOURCE_REMOVE;
+	if (!info->resolv && info->resolv_id) {
+		DBG("resolv already removed");
+		return;
 	}
 
-	/* If this fails after being connected it means configuration error
-	 * that results in connection errors. */
-	err = parse_endpoint(info->endpoint_fqdn,
-			info->port, &addr);
+	info->remove_resolv_id = g_timeout_add(0, remove_resolv, info);
+	info->resolv_id = 0;
```

Comment:
this looks suspicious. why is resolv_id just cleared here? surely if it is still active it should be canceled / cleared elsewhere / something?
or alternatively: makes me think whether that "resolv already removed" condition a bit earlier is correct or not?

Reply:
Yes this needs clarification. The thing is that when the GResolv callback is returned, so the lookup is completed, GResolv cannot be unref'd within the callback. Therefore, the timeout function to call `remove_resolv()` is used, by zeroing the resolv_id so the lookup is not attempted to be canceled - as it completed. The resolv struct is manipulated by gresolv.c after the callback returns, it would segfault if done here.

A bit hacky, needs a better approach? Yes to both but I can live with this approach.

Reply:
A bit hacky, yes. But with explanation understandable. I guess it is ok.